File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int n, int k, vector<int> enemy) {
    if (k >= enemy.size()) return enemy.size();
    priority_queue<int, vector<int>, greater<int>> pq;
    int total = 0, queSum = 0;
    for (int i = 0; i < enemy.size(); i++) {
        pq.push(enemy[i]);
        total += enemy[i];
        queSum += enemy[i];
        if (pq.size() > k) {
            queSum -= pq.top();
            pq.pop();
        }
        if (total - queSum > n) return i;
    }
    return enemy.size();
}

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.java
import java.util.*;

class Solution {
    public int solution(int n, int k, int[] enemy) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int i = 0; i < enemy.length; i++) {
            if (pq.size() < k) pq.add(enemy[i]);
            else {
                pq.add(enemy[i]);
                int curr = pq.poll();
                if (n >= curr) n -= curr;
                else return i;
            }
        }
        return enemy.length;
    }
}